src/eigsep_sensors/imu/imu.py

```
@@ -0,0 +1,271 @@
+from abc import ABC, abstractmethod
+import serial
+import time
+import numpy as np
+
+class IMU(ABC):
+    """
+    Base class for IMU (Inertial Measurement Unit) sensors.
+
+    Provides a generic interface for requesting and reading data from IMUs
+    that communicate over a serial connection. Child classes should implement
+    the `_parse_line` method if they use serial-based communication.
+    """
+
+    def __init__(self, port, timeout=1):
+        """
+        Initialize the BNO085 sensor over a serial connection.
+
+        Args:
+            port (str): Serial port to which the sensor is connected.
+            timeout (int or float): Read timeout in seconds.
+        """
+        self.ser = serial.Serial(port=port, baudrate=115200, timeout=timeout)
+
+    def _request_imu(self):
+        """
+        Send a 'REQ' signal over serial to request sensor data.
+        """
+        self.ser.write(b"REQ\n")
+
+    def read_imu(self):
+        """
+        Request and read data from the IMU via serial.
+
+        Returns:
+            dict or None: Parsed data as a dictionary if successful,
+                          otherwise None if no data is returned.
+        """
+        self._request_imu()
+        response = self.ser.readline().decode().strip()
+        if not response:
+            return None
+        return self._parse_line(response)
+    
+    @abstractmethod
+    def _parse_line(self, line):
+        """
+        Filled in subclasses.
+        """
+        pass
+
+    def close(self):
+        """
+        Closes serial connection.
+        """
+        self.ser.close()
+
+    
+class IMU_MMA8451(IMU):
+    """
+    IMU class for the Adafruit MMA8451 accelerometer sensor over I2C.
+
+    Inherits from IMU, but uses I2C instead of serial. This class initializes
+    the I2C bus and sets up the sensor for acceleration measurements.
+    """
+
+    def _parse_line(self, line):

```

Here, you try/except the whole function. It is more explicit to in this case just let an exception be raised if there is one and make downstream users handle it. The risk of doing it the way it's currently implemented is that real errors might be hidden and made hard to debug.